pkg/proc/gdbserial/undo.go

```diff
@@ -2,19 +2,318 @@ package gdbserial
 
 import (
 	"bytes"
+	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"os/exec"
+	"os/user"
 	"path/filepath"
 	"runtime"
+	"sort"
 	"strconv"
 	"strings"
 	"syscall"
 
 	"github.com/undoio/delve/pkg/proc"
 )
 
+// State relating to an Undo "session" - this is runtime state that corresponds with.
+type undoSession struct {
+	checkpointLastId int
+	checkpoints      map[int]proc.Checkpoint
+}
+
+// Create a new undoSession structure, which captures the runtime state relating to an on-disk
+// "session" file - currently only bookmarks are supported.
+func newUndoSession() *undoSession {
+	return &undoSession{
+		checkpointLastId: 1,
+		checkpoints:      make(map[int]proc.Checkpoint),
+	}
+}
+
+// Validate a checkpoint note to ensure easy interopability with UDB bookmarks.
```

Comment:
Consider documenting the result, i.e. `nil` if the checkpoint note is valid, or an error if not.